src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs

```diff
@@ -9,61 +9,63 @@
 // Redistribution and use in source and binary forms with or without
 // modifications are permitted.
 
-using Neo.IO.Data.LevelDB;
+using Neo.IO.Storage.LevelDB;
 using Neo.Persistence;
 using System.Collections.Generic;
-using LSnapshot = Neo.IO.Data.LevelDB.Snapshot;
 
 namespace Neo.Plugins.Storage
 {
     internal class Snapshot : ISnapshot
     {
-        private readonly DB db;
-        private readonly LSnapshot snapshot;
-        private readonly ReadOptions options;
-        private readonly WriteBatch batch;
+        private readonly DB _db;
+        private readonly SnapShot _snapshot;
+        private readonly WriteBatch _batch;
+        private readonly ReadOptions _readOptions;
+
+        private readonly object _lock = new();
 
         public Snapshot(DB db)
         {
-            this.db = db;
-            snapshot = db.GetSnapshot();
-            options = new ReadOptions { FillCache = false, Snapshot = snapshot };
-            batch = new WriteBatch();
+            _db = db;
+            _snapshot = db.CreateSnapshot();
+            _batch = new();
+            _readOptions = new ReadOptions { FillCache = false, Snapshot = _snapshot };
         }
 
         public void Commit()
         {
-            db.Write(WriteOptions.Default, batch);
+            lock (_lock)
+                _db.Write(_batch);
         }
 
         public void Delete(byte[] key)
         {
-            batch.Delete(key);
+            lock (_lock)
+                _batch.Delete(key);
         }
 
         public void Dispose()
         {
-            snapshot.Dispose();
-        }
-
-        public IEnumerable<(byte[] Key, byte[] Value)> Seek(byte[] prefix, SeekDirection direction = SeekDirection.Forward)
-        {
-            return db.Seek(options, prefix, direction, (k, v) => (k, v));
+            _snapshot.Dispose();
+            _batch.Dispose();
+            _readOptions.Dispose();
         }
 
         public void Put(byte[] key, byte[] value)
         {
-            batch.Put(key, value);
+            lock (_lock)
```

Comment:
@Jim8y already introduced something to ensure that is not used in non-thread safe cases

Reply:
Still best to do this. All of leveldb is thread safe, except `writebatch`, isn't thread safe. Better to do it at the source. You can use leveldb in multi-threads. Just not `writebatch`.

Reply:
@Jim8y do you think that we still need this lock?

Reply:
> @Jim8y do you think that we still need this lock?

I would support to have the lock here, reason is i prefer to have `module level security and safety`, instead of `project level security and safety`, when we have a module, we should ensure the module always work fine (no unexpected or undefined behavior) no matter it's being used the right way or the wrong way.

Reply:
Then, we need a read and write lock

Reply:
That's what the lock currently does is to read and write lock.

Reply:
> Lock isn't required for reading in level DB.

Why don't?

Reply:
> > Lock isn't required for reading in level DB.
> 
> Why don't?

Will limit performance. Level Db says the lock is only required when writing to a batch